packages/types/src/codec/utils/sortValues.ts

```diff
@@ -0,0 +1,53 @@
+// Copyright 2017-2021 @polkadot/types authors & contributors
+// SPDX-License-Identifier: Apache-2.0
+
+import type { Codec } from '../../types';
+
+import { AbstractArray } from '../AbstractArray';
+import { AbstractInt } from '../AbstractInt';
+import { Enum } from '../Enum';
+import { Struct } from '../Struct';
+
+/**
+* Sort keys/values of BTreeSet/BTreeMap in acending order for encoding compatibility with Rust's BTreeSet/BTreeMap
+* (https://doc.rust-lang.org/stable/std/collections/struct.BTreeSet.html)
+* (https://doc.rust-lang.org/stable/std/collections/struct.BTreeMap.html)
+*/
+export function sortAsc<V extends Codec | Codec[] | number[] | number | Uint8Array = Codec> (a: V, b: V): number {
+  if (typeof a === 'number') {
+    return a - (b as typeof a);
+  } else if (a instanceof Struct) {
+    return sortAsc(Array.from(a.values()), Array.from((b as Struct).values()));
+  } else if (a instanceof Enum) {
+    return sortAsc(a.index, (b as Enum).index) || sortAsc(a.value, (b as Enum).value);
+  } else if (a instanceof AbstractArray || a instanceof Uint8Array || Array.isArray(a)) {
+    // Vec, Tuple, Bytes etc.
+    let sortRes = 0; const lenA = a.length; const lenB = (b as typeof a).length;
+    const minLen = Math.min(lenA, lenB);
+
+    for (let i = 0; i < minLen; ++i) {
+      sortRes = sortAsc(a[i], (b as typeof a)[i]);
+
+      if (sortRes !== 0) {
+        return sortRes;
+      }
+    }
+
+    return lenA - lenB;
+  } else if (a instanceof AbstractInt) {
+    return a.eq(b) ? 0 : a.lt(b as typeof a) ? -1 : 1;
+  } else if (a && typeof a.toU8a === 'function') {
+    // Fallback (Bytes, Text)
+    return sortAsc(a.toU8a(true), (b as Codec).toU8a(true));
+  } else {
+    throw new Error(`Attempting to sort unrecognized value: ${JSON.stringify(a)}`);
+  }
+}
+
+export function sortSet<V extends Codec = Codec> (set: Set<V>): Set<V> {
+  return new Set(Array.from(set).sort(sortAsc));
+}
+
+export function sortMap<K extends Codec = Codec, V extends Codec = Codec> (map: Map<K, V>): Map<K, V> {
+  return new Map(Array.from(map.entries()).sort(([keyA], [keyB]) => sortAsc(keyA, keyB)));
```

Comment:
Really wondering if this is not just a 

```js
return new Map(Array.from(map.entries()).sort(([keyA], [keyB]) => u8aCmp(keyA.toU8a(), keyB.toU8a())));
```

Or at least for most of them - the Codec numbers are the only issue since they are LE encoded.

Reply:
Unfortunately it's more complicated.

For example a `Struct` with 2 `Text` fields should be sorted by first filed first and then by second field only in case first fields `a` and `b` are equal.

The example below shows that it's not possible to achieve this with `u8aCmp` (expected sorting is: `{ handle: 'a' }`, `{ handle: 'aa' }`, `{ handle: 'b' }`)

![image](https://user-images.githubusercontent.com/19978260/126828702-8c9dab1a-8271-4289-8e9b-d4423189187d.png)

Then there is also an issue with nested types, like ints inside tuples inside structs etc.

Reply:
100%, was wondering about the `Vec<T>`-like encoding where the length is encoded, makes sense that it is raw data only.